<a href="https://colab.research.google.com/github/VitalyGladyshev/gb_nlp/blob/main/HW_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ДЗ 02 Пока заглушка. Скоро сделаю и сдам...

In [ ]:
import numpy as np
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [3]:
path_nlp = "/content/gdrive/My Drive/Colab Notebooks/NLP/"

In [16]:
train_cor = pd.read_pickle(path_nlp + "train_cor.pkl")
test_cor = pd.read_pickle(path_nlp + "test_cor.pkl")

In [17]:
train_cor.tail(2)

id  ...                                   tweet_lemmatized
31960  31961  ...  {vandalise, wso, calgary, act, temple, sikh, c...
31961  31962  ...                                    {thank, follow}

[2 rows x 8 columns]

In [18]:
test_cor.tail(2)

id  ...                                   tweet_lemmatized
49157  49158  ...  {conference, lead, cultureofdevelopment, happy...
49158  49159  ...  {newsong, download, glad, free, newmusic, shoe...

[2 rows x 7 columns]

In [19]:
test_cor['label'] = -1

In [20]:
data = train_cor.copy()
data = data.append(test_cor, ignore_index=True)
data.tail(2)

id  ...                                   tweet_lemmatized
49157  49158  ...  {conference, lead, cultureofdevelopment, happy...
49158  49159  ...  {newsong, download, glad, free, newmusic, shoe...

[2 rows x 8 columns]

## Задание 1

Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
- Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
- Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
- Исключим стоп-слова с помощью stop_words='english'.
- Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().


## Задание 2

Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
- Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
- Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
- Исключим стоп-слова с помощью stop_words='english'.
- Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().


## Задание 3

Натренируем gensim.models.Word2Vec модель на наших данных.
- Тренировать будем на токенизированных твитах combine_df['tweet_token']
- Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
- Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.


## Задание 4

Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

## Задание 5

Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
<br>Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

## Задание 6

Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
<br>Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size)) и поделить финальный вектор на количество слов в твите. На выходе должен получиться wordvec_df.shape = (49159, 200).
